auth/realm/base/src/main/java/org/wildfly/security/auth/realm/FileSystemSecurityRealm.java

```diff
@@ -135,22 +218,46 @@
      * @param root the root path of the identity store
      * @param nameRewriter the name rewriter to apply to looked up names
      * @param levels the number of levels of directory hashing to apply
-     * @param encoded whether identity names should be BASE32 encoded before using as filename
+     * @param encoded whether identity names should be BASE32 encoded before using as filename (only applies if the security realm is unencrypted)
      * @param hashCharset the character set to use when converting password strings to a byte array. Uses UTF-8 by default.
      * @param hashEncoding the string format for the hashed passwords. Uses Base64 by default.
+     * @param providers The providers supplier
+     * @param secretKey the SecretKey used to encrypt and decrypt the security realm (if {@code null}, the security realm will be unencrypted)
+     * @param privateKey the PrivateKey used for writing the signature of the identities in security realm (if {@code null}, the security realm will not have integrity enabled)
+     * @param publicKey the PublicKey used for verifying the signature of the identities in security realm (if {@code null}, the security realm will not have integrity enabled)
      */
-    public FileSystemSecurityRealm(final Path root, final NameRewriter nameRewriter, final int levels, final boolean encoded, final Encoding hashEncoding, final Charset hashCharset, Supplier<Provider[]> providers) {
+    public FileSystemSecurityRealm(final Path root, final NameRewriter nameRewriter, final int levels, final boolean encoded, final Encoding hashEncoding, final Charset hashCharset, Supplier<Provider[]> providers, final SecretKey secretKey, final PrivateKey privateKey, final PublicKey publicKey) {
         final SecurityManager sm = System.getSecurityManager();
         if (sm != null) {
             sm.checkPermission(CREATE_SECURITY_REALM);
         }
         this.root = root;
         this.nameRewriter = nameRewriter;
         this.levels = levels;
-        this.encoded = encoded;
+        this.encoded = secretKey == null && encoded;
         this.hashCharset = hashCharset != null ? hashCharset : StandardCharsets.UTF_8;
         this.hashEncoding = hashEncoding != null ? hashEncoding : Encoding.BASE64;
-        this.providers = providers;
+        this.providers = providers != null ? providers : INSTALLED_PROVIDERS;
+        this.secretKey = secretKey;
+        this.privateKey = publicKey != null ? privateKey : null;
+        this.publicKey = privateKey != null ? publicKey : null;
```

Comment:
Once the constructor is package private, it should only ever be called with valid params (since only the build method and the other constructors can call it). So you wouldn't need these checks here.

Reply:
Moved any checks into the FileSystemSecurityRealmBuilder